chain/client/src/stateless_validation/validate.rs

```diff
@@ -0,0 +1,180 @@
+use super::partial_witness::witness_part_length;
+use itertools::Itertools;
+use near_chain::types::Tip;
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::stateless_validation::chunk_endorsement::ChunkEndorsementV2;
+use near_primitives::stateless_validation::partial_witness::{
+    PartialEncodedStateWitness, MAX_COMPRESSED_STATE_WITNESS_SIZE,
+};
+use near_primitives::stateless_validation::ChunkProductionKey;
+use near_primitives::types::{AccountId, BlockHeightDelta};
+use near_primitives::validator_signer::ValidatorSigner;
+use near_store::{DBCol, Store, FINAL_HEAD_KEY, HEAD_KEY};
+
+/// This is taken to be the same value as near_chunks::chunk_cache::MAX_HEIGHTS_AHEAD, and we
+/// reject partial witnesses with height more than this value above the height of our current HEAD
+const MAX_HEIGHTS_AHEAD: BlockHeightDelta = 5;
+
+/// Function to validate the partial encoded state witness. In addition of ChunkProductionKey, we check the following:
+/// - part_ord is valid and within range of the number of expected parts for this chunk
+/// - partial_witness signature is valid and from the expected chunk_producer
+/// TODO(stateless_validation): Include checks from handle_orphan_state_witness in orphan_witness_handling.rs
+/// These include checks based on epoch_id validity, witness size, height_created, distance from chain head, etc.
+pub fn validate_partial_encoded_state_witness(
+    epoch_manager: &dyn EpochManagerAdapter,
+    partial_witness: &PartialEncodedStateWitness,
+    signer: &ValidatorSigner,
+    store: &Store,
+) -> Result<bool, Error> {
+    let ChunkProductionKey { shard_id, epoch_id, height_created } =
+        partial_witness.chunk_production_key();
+    let num_parts =
+        epoch_manager.get_chunk_validator_assignments(&epoch_id, shard_id, height_created)?.len();
+    if partial_witness.part_ord() >= num_parts {
+        return Err(Error::InvalidPartialChunkStateWitness(format!(
+            "Invalid part_ord in PartialEncodedStateWitness: {}",
+            partial_witness.part_ord()
+        )));
+    }
+
+    let max_part_len =
+        witness_part_length(MAX_COMPRESSED_STATE_WITNESS_SIZE.as_u64() as usize, num_parts);
+    if partial_witness.part_size() > max_part_len {
+        return Err(Error::InvalidPartialChunkStateWitness(format!(
+            "Part size {} exceed limit of {} (total parts: {})",
+            partial_witness.part_size(),
+            max_part_len,
+            num_parts
+        )));
+    }
+
+    if !validate_chunk_production_key(
```

Comment:
Separated the checks between chunk_production_key and rest of the partial witness. Reused the logic for `validate_chunk_endorsement`
Please compare with the original `validate_partial_encoded_state_witness` for review